File: src/main/java/com/github/morningzeng/toolset/ui/crypto/AbstractCryptoComponent.java
package com.github.morningzeng.toolset.ui.crypto;

import com.github.morningzeng.toolset.component.LanguageTextArea;
import com.github.morningzeng.toolset.model.Children;
import com.github.morningzeng.toolset.ui.AbstractCryptoPropComponent;
import com.intellij.openapi.fileTypes.PlainTextLanguage;
import com.intellij.openapi.project.Project;
import lombok.extern.slf4j.Slf4j;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * @author Morning Zeng
 * @since 2024-07-09
 */
@Slf4j
sealed abstract class AbstractCryptoComponent<T extends Children<T>> extends AbstractCryptoPropComponent<T> permits AbstractSymmetricCryptoComponent, AsymmetricComponent, HashComponent {
    protected final LanguageTextArea encryptArea;
    protected final LanguageTextArea decryptArea;

    AbstractCryptoComponent(final Project project) {
        super(project);
        this.encryptArea = new LanguageTextArea(PlainTextLanguage.INSTANCE, project, "");
        this.decryptArea = new LanguageTextArea(PlainTextLanguage.INSTANCE, project, "");
        this.encryptArea.setPlaceholder("Encrypted text content");
        this.decryptArea.setPlaceholder("Decrypted text content");
    }

    @Override
    protected Stream<T> flatProps(final List<T> props) {
        return props.stream()
                .filter(this.filterProp())
                .sorted(Children.comparable())
                .mapMulti((prop, consumer) -> {
                    consumer.accept(prop);
                    Optional.ofNullable(prop.getChildren()).ifPresent(
                            ts -> ts.stream()
                                    .filter(this.filterProp())
                                    .sorted(Children.comparable())
                                    .forEach(consumer)
                    );
                });
    }

}

File: src/main/kotlin/com/github/morningzeng/toolbox/ui/view/crypto/AbstractCryptoComponent.kt
package com.github.morningzeng.toolbox.ui.view.crypto

import com.github.morningzeng.toolbox.model.Children
import com.github.morningzeng.toolbox.ui.action.AutoFormatAction
import com.github.morningzeng.toolbox.ui.action.HistoryAction
import com.github.morningzeng.toolbox.ui.component.LanguageTextArea
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.project.Project
import java.util.stream.Stream

/**
 * @author Morning Zeng
 * @since 2025-05-19
 */
abstract class AbstractCryptoComponent<T : Children<T>>(
    project: Project,
) : AbstractCryptoPropComponent<T>(project) {

    private val regex = Regex("(?<=^.{10}).+?(?=.{10}$)")
    var placeholderString = " ...... "

    protected val encHistoryAction = HistoryAction<String>().apply {
        historyRender { it.replace(regex, placeholderString) }
    }
    protected val decHistoryAction = HistoryAction<String>().apply {
        historyRender { it.replace(regex, placeholderString) }
    }

    protected val encryptArea: LanguageTextArea = object : LanguageTextArea(project) {
        override fun defaultRightBarActions(): Array<AnAction> {
            return arrayOf(
                softWrapAction(),
                scrollToEndAction(),
                encHistoryAction.also {
                    it.setTextArea(this)
                    it.textAreaEvent(textConvert(), tConvert())
                },
                clearAllAction()
            )
        }
    }
        .also { it.setPlaceholder("Encrypted text content") }
    protected val decryptArea: LanguageTextArea = object : LanguageTextArea(project) {
        override fun defaultRightBarActions(): Array<AnAction> {
            return arrayOf(
                softWrapAction(),
                scrollToEndAction(),
                decHistoryAction.also {
                    it.setTextArea(this)
                    it.textAreaEvent(textConvert(), tConvert())
                },
                AutoFormatAction(this),
                clearAllAction()
            )
        }
    }
        .also { it.setPlaceholder("Decrypted text content") }

    override fun flatProps(props: MutableList<T>?): Stream<T> {
        return props?.stream()
            ?.sorted(Children.comparable())
            ?.mapMulti { t, consumer ->
                consumer.accept(t)
                t.children.stream()
                    .filter(filterProp())
                    .sorted(Children.comparable())
                    .forEach { consumer.accept(it) }
            }
            ?: Stream.empty()
    }

    open fun textConvert(): (String) -> String {
        return { it }
    }

    open fun tConvert(): (String) -> String? {
        return { it }
    }

}